docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,893 @@
+before the configuration (which defines parameters like `logger.log_level`) is fully resolved. Keeping configuration
+tracing as a straightforward on-off switch, without the complexity of log levels, further simplifies the user's
+interaction and guarantees that all relevant trace data is captured when enabled.
+
+#### Sensitive Information
+
+While the trace provides detailed information, caution should be taken to avoid exposing sensitive data, such as private
+keys, ensuring security isn't compromised.
+
+#### Further Integration
+
+Incorporating these traces into configuration-related error messages can further streamline troubleshooting, enhancing
+user experience. (Refer to: [Proposal 6 - Exhaustive Error Messages](#proposal-6---exhaustive-error-messages))
+
+#### Summary
+
+This proposal aims to allow users to see how the config gets pieced together, making it easier to set up and fix any
+hiccups along the way.
+
+#### See Also
+
+- [[suggestion] Trace configuration parameters resolution · Issue #3502 · hyperledger/iroha](https://github.com/hyperledger/iroha/issues/3502)
+
+### Proposal 8 - Dual-View Configuration Resolution
+
+The dichotomy between configurations which are user-friendly and those that are efficient for software interpretation
+and validation can be bridged through a dual-view configuration system: "User Config" and "Resolved Config".
+
+#### Objective
+
+Provide a distinction between the configuration as perceived by the user and the software's operational, validated
+configuration view, striking a balance between user-friendliness and software efficiency & correctness.
+
+#### Rationale
+
+1. **User Clarity and Software Efficiency:** Keeping a representation that directly mirrors the reference for users
+   (User Config) and a separate, robustly-typed structure for the software's operational use (Resolved Config) benefits
+   both users and developers.
+2. **Parse, Not Validate:** Transitioning from the User Config to the Resolved Config involves rigorous parsing and
+   validation. Upon the formation of the Resolved Config, it's inherently valid, minimising the likelihood of runtime
+   configuration errors.
+3. **Domain-Specific Considerations:** The transformation phase applies domain (in this case, Iroha) specific rules and
+   logic, ensuring the final Resolved Config is both syntactically and semantically apt.
+
+#### Research about Resolved Configuration
+
+##### Redundant Fields
+
+In the background about [redundant fields](#redundant-fields), it's mentioned that these fields were made available to
+end users primarily because the user config wasn't separated from the resolved config. However, in a two-stage
+configuration resolution, these redundant fields can be managed more efficiently.
+
+##### Root-Level `public_key` and `private_key`
+
+When users provide their configuration input, it should be possible for them to specify only the private key. From this,
```

Comment:
if we implement this logic we might just drop providing `public_key` from the configuration whatsoever

Reply:
Keys configuration has a few issues:

- #2135
- #2824
- #3240

I am not sure what of these should be included into the RFC. Thus, I chose to stick with the status quo.

Let me put a note somewhere in the RFC about it and mention these issues.

Reply:
Included this note as a footnote.